pkg/applicationserver/applicationserver.go

```diff
@@ -621,31 +828,31 @@ func (as *ApplicationServer) handleUplink(ctx context.Context, ids ttnpb.EndDevi
 					dev.DevAddr = ids.DevAddr
 					mask = append(mask, "session", "ids.dev_addr")
 					logger.Debug("Restored session")
-				}
-				dev.PendingSession = nil
-				mask = append(mask, "pending_session")
 
-				// At this point, the application downlink queue in the Network Server is invalid; recalculation is necessary.
-				// Next AFCntDown 1 is assumed. If this is a LoRaWAN 1.0.x end device and the Network Server sent MAC layer
-				// downlink already, the Network Server will trigger the DownlinkQueueInvalidated event. Therefore, this
-				// recalculation may result in another recalculation.
-				client := ttnpb.NewAsNsClient(link.conn)
-				res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
-				if err != nil {
-					log.WithError(err).Warn("Failed to list downlink queue for recalculation; clearing the downlink queue")
-					req := &ttnpb.DownlinkQueueRequest{
-						EndDeviceIdentifiers: ids,
-					}
-					_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
+					// At this point, the application downlink queue in the Network Server is invalid; recalculation is necessary.
+					// Next AFCntDown 1 is assumed. If this is a LoRaWAN 1.0.x end device and the Network Server sent MAC layer
+					// downlink already, the Network Server will trigger the DownlinkQueueInvalidated event. Therefore, this
+					// recalculation may result in another recalculation.
+					client := ttnpb.NewAsNsClient(link.conn)
+					res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
 					if err != nil {
-						log.WithError(err).Warn("Failed to clear the downlink queue; any queued items in the Network Server are invalid")
-						events.Publish(evtInvalidQueueDataDown(ctx, ids, err))
-					} else {
-						events.Publish(evtLostQueueDataDown(ctx, ids, err))
+						logger.WithError(err).Warn("Failed to list downlink queue for recalculation; clearing the downlink queue")
+						req := &ttnpb.DownlinkQueueRequest{
+							EndDeviceIdentifiers: ids,
+						}
+						_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
+						if err != nil {
+							logger.WithError(err).Warn("Failed to clear the downlink queue; any queued items in the Network Server are invalid")
+							events.Publish(evtInvalidQueueDataDown(ctx, ids, err))
+						} else {
+							events.Publish(evtLostQueueDataDown(ctx, ids, err))
+						}
+					} else if err := as.recalculateDownlinkQueue(ctx, dev, link, previousSession, res.Downlinks, 1); err != nil {
+						logger.WithError(err).Warn("Failed to recalculate downlink queue; items lost")
 					}
-				} else if err := as.recalculateDownlinkQueue(ctx, dev, previousSession, res.Downlinks, 1, link); err != nil {
-					log.WithError(err).Warn("Failed to recalculate downlink queue")
 				}
+				dev.PendingSession = nil
+				mask = append(mask, "pending_session")
 			} else if dev.Session.AppSKey == nil {
 				return nil, nil, errNoAppSKey.New()
 			}
```

Comment:
At this point `dev.Session` must be non-`nil` and `AppSKey` being `nil` must cause an error in pending session switch as well (https://github.com/TheThingsNetwork/lorawan-stack/pull/2163/files#diff-2719373b9eca80ef1abcf09309389399R810)
```suggestion
			}
			if dev.Session.AppSKey == nil {
				return nil, nil, errNoAppSKey.New()
			}
```